File: src/validator.py
"""Validation and safety checks for embroidery file resizing."""

from enum import Enum
from typing import Tuple, Optional
from .utils import calculate_resize_percentage


class ValidationLevel(Enum):
    """Validation severity levels."""
    SAFE = "safe"
    WARNING = "warning"
    DANGER = "danger"
    CRITICAL = "critical"


class ValidationResult:
    """Result of a validation check."""

    def __init__(self, level: ValidationLevel, message: str, can_proceed: bool = True):
        self.level = level
        self.message = message
        self.can_proceed = can_proceed

    def __str__(self):
        return f"[{self.level.value.upper()}] {self.message}"


class ResizeValidator:
    """Validator for embroidery resize operations."""

    # Safe resize limits (percentage)
    SAFE_RESIZE_LIMIT = 20.0
    WARNING_RESIZE_LIMIT = 30.0
    CRITICAL_RESIZE_LIMIT = 50.0

    # Stitch density limits (mm)
    MIN_STITCH_DENSITY = 0.2  # Too dense, may break needles
    OPTIMAL_STITCH_DENSITY_MIN = 0.4
    OPTIMAL_STITCH_DENSITY_MAX = 0.45
    MAX_STITCH_DENSITY = 1.0  # Too sparse, poor quality

    def __init__(self):
        self.results = []

    def validate_resize_percentage(
        self, original_size: float, new_size: float
    ) -> ValidationResult:
        """
        Validate the resize percentage is within safe limits.

        Args:
            original_size: Original dimension
            new_size: New dimension

        Returns:
            ValidationResult
        """
        percent_change = abs(calculate_resize_percentage(original_size, new_size))

        if percent_change <= self.SAFE_RESIZE_LIMIT:
            return ValidationResult(
                ValidationLevel.SAFE,
                f"Resize of {percent_change:.1f}% is within safe limits (±{self.SAFE_RESIZE_LIMIT}%)",
            )
        elif percent_change <= self.WARNING_RESIZE_LIMIT:
            return ValidationResult(
                ValidationLevel.WARNING,
                f"Resize of {percent_change:.1f}% may affect quality. Consider staying within ±{self.SAFE_RESIZE_LIMIT}%",
            )
        elif percent_change <= self.CRITICAL_RESIZE_LIMIT:
            return ValidationResult(
                ValidationLevel.DANGER,
                f"Resize of {percent_change:.1f}% is significant and may cause problems. Re-digitizing recommended for best results",
            )
        else:
            return ValidationResult(
                ValidationLevel.CRITICAL,
                f"Resize of {percent_change:.1f}% is too extreme. Original design quality will be severely compromised",
                can_proceed=False,
            )

    def validate_stitch_density(self, density: float) -> ValidationResult:
        """
        Validate stitch density is within acceptable range.

        Args:
            density: Average stitch density in mm

        Returns:
            ValidationResult
        """
        if density < self.MIN_STITCH_DENSITY:
            return ValidationResult(
                ValidationLevel.CRITICAL,
                f"Stitch density ({density:.3f}mm) is too dense. May cause needle breaks or fabric damage",
                can_proceed=False,
            )
        elif density < self.OPTIMAL_STITCH_DENSITY_MIN:
            return ValidationResult(
                ValidationLevel.WARNING,
                f"Stitch density ({density:.3f}mm) is denser than optimal ({self.OPTIMAL_STITCH_DENSITY_MIN}-{self.OPTIMAL_STITCH_DENSITY_MAX}mm)",
            )
        elif density <= self.OPTIMAL_STITCH_DENSITY_MAX:
            return ValidationResult(
                ValidationLevel.SAFE,
                f"Stitch density ({density:.3f}mm) is optimal",
            )
        elif density <= self.MAX_STITCH_DENSITY:
            return ValidationResult(
                ValidationLevel.WARNING,
                f"Stitch density ({density:.3f}mm) is sparser than optimal ({self.OPTIMAL_STITCH_DENSITY_MIN}-{self.OPTIMAL_STITCH_DENSITY_MAX}mm)",
            )
        else:
            return ValidationResult(
                ValidationLevel.DANGER,
                f"Stitch density ({density:.3f}mm) is too sparse. Quality will be poor",
            )

    def validate_dimensions(
        self, width: float, height: float, max_width: Optional[float] = None, max_height: Optional[float] = None
    ) -> ValidationResult:
        """
        Validate pattern dimensions fit within constraints.

        Args:
            width: Pattern width in mm
            height: Pattern height in mm
            max_width: Maximum allowed width in mm (optional)
            max_height: Maximum allowed height in mm (optional)

        Returns:
            ValidationResult
        """
        if max_width and width > max_width:
            return ValidationResult(
                ValidationLevel.CRITICAL,
                f"Width ({width:.1f}mm) exceeds maximum ({max_width:.1f}mm)",
                can_proceed=False,
            )

        if max_height and height > max_height:
            return ValidationResult(
                ValidationLevel.CRITICAL,
                f"Height ({height:.1f}mm) exceeds maximum ({max_height:.1f}mm)",
                can_proceed=False,
            )

        return ValidationResult(
            ValidationLevel.SAFE,
            f"Dimensions ({width:.1f}mm × {height:.1f}mm) are valid",
        )

    def validate_all(
        self,
        original_width: float,
        original_height: float,
        new_width: float,
        new_height: float,
        stitch_density: Optional[float] = None,
    ) -> Tuple[bool, list]:
        """
        Run all validations.

        Args:
            original_width: Original width in mm
            original_height: Original height in mm
            new_width: New width in mm
            new_height: New height in mm
            stitch_density: Stitch density in mm (optional)

        Returns:
            Tuple of (can_proceed, list of ValidationResults)
        """
        results = []

        # Check width resize
        width_result = self.validate_resize_percentage(original_width, new_width)
        results.append(width_result)

        # Check height resize
        height_result = self.validate_resize_percentage(original_height, new_height)
        results.append(height_result)

        # Check stitch density if provided
        if stitch_density is not None:
            density_result = self.validate_stitch_density(stitch_density)
            results.append(density_result)

        # Determine if we can proceed
        can_proceed = all(r.can_proceed for r in results)

        return can_proceed, results

File: simpleskale-v4/src/engine/validator.ts
/**
 * SimpleSkale 4.0 - Validation Engine
 *
 * Ported from Python validator.py with equivalent functionality.
 * Validates resize operations for safety and quality.
 */

import type { MachineProfile, ValidationLevel, ValidationResult } from "../types";
import { ValidationLevel as VLevel } from "../types";
import { calculateResizePercentage } from "../utils/calculations";

/**
 * Validation thresholds (ported from Python ResizeValidator)
 */
export const VALIDATION_THRESHOLDS = {
  SAFE_RESIZE_LIMIT: 20.0,
  WARNING_RESIZE_LIMIT: 30.0,
  CRITICAL_RESIZE_LIMIT: 50.0,
  MIN_STITCH_DENSITY: 0.2, // Too dense, may break needles
  OPTIMAL_DENSITY_MIN: 0.4,
  OPTIMAL_DENSITY_MAX: 0.45,
  MAX_STITCH_DENSITY: 1.0, // Too sparse, poor quality
} as const;

/**
 * Create a validation result
 */
function createValidationResult(
  level: ValidationLevel,
  message: string,
  canProceed: boolean = true,
  details?: string
): ValidationResult {
  return { level, message, canProceed, details };
}

/**
 * Validate resize percentage is within safe limits
 *
 * Ported from Python validate_resize_percentage()
 */
export function validateResizePercentage(
  originalSize: number,
  newSize: number
): ValidationResult {
  const percentChange = Math.abs(calculateResizePercentage(originalSize, newSize));

  if (percentChange <= VALIDATION_THRESHOLDS.SAFE_RESIZE_LIMIT) {
    return createValidationResult(
      VLevel.SAFE,
      `Resize of ${percentChange.toFixed(1)}% is within safe limits (±${VALIDATION_THRESHOLDS.SAFE_RESIZE_LIMIT}%)`
    );
  } else if (percentChange <= VALIDATION_THRESHOLDS.WARNING_RESIZE_LIMIT) {
    return createValidationResult(
      VLevel.WARNING,
      `Resize of ${percentChange.toFixed(1)}% may affect quality. Consider staying within ±${VALIDATION_THRESHOLDS.SAFE_RESIZE_LIMIT}%`,
      true,
      "This resize is outside the recommended safe range but should still produce acceptable results for most designs."
    );
  } else if (percentChange <= VALIDATION_THRESHOLDS.CRITICAL_RESIZE_LIMIT) {
    return createValidationResult(
      VLevel.DANGER,
      `Resize of ${percentChange.toFixed(1)}% is significant and may cause problems. Re-digitizing recommended for best results`,
      true,
      "At this scale, you may experience quality issues. Consider having the design professionally re-digitized."
    );
  } else {
    return createValidationResult(
      VLevel.CRITICAL,
      `Resize of ${percentChange.toFixed(1)}% is too extreme. Original design quality will be severely compromised`,
      false,
      "This resize is far outside safe limits. The result will likely be unusable. Professional re-digitizing is strongly recommended."
    );
  }
}

/**
 * Validate stitch density is within acceptable range
 *
 * Ported from Python validate_stitch_density()
 */
export function validateStitchDensity(densityMm: number): ValidationResult {
  if (densityMm < VALIDATION_THRESHOLDS.MIN_STITCH_DENSITY) {
    return createValidationResult(
      VLevel.CRITICAL,
      `Stitch density (${densityMm.toFixed(3)}mm) is too dense. May cause needle breaks or fabric damage`,
      false,
      "Stitches are too close together. This can break needles, damage fabric, and cause thread bunching."
    );
  } else if (densityMm < VALIDATION_THRESHOLDS.OPTIMAL_DENSITY_MIN) {
    return createValidationResult(
      VLevel.WARNING,
      `Stitch density (${densityMm.toFixed(3)}mm) is denser than optimal (${VALIDATION_THRESHOLDS.OPTIMAL_DENSITY_MIN}-${VALIDATION_THRESHOLDS.OPTIMAL_DENSITY_MAX}mm)`,
      true,
      "This density is higher than recommended but may work depending on fabric and thread type."
    );
  } else if (densityMm <= VALIDATION_THRESHOLDS.OPTIMAL_DENSITY_MAX) {
    return createValidationResult(
      VLevel.SAFE,
      `Stitch density (${densityMm.toFixed(3)}mm) is optimal`
    );
  } else if (densityMm <= VALIDATION_THRESHOLDS.MAX_STITCH_DENSITY) {
    return createValidationResult(
      VLevel.WARNING,
      `Stitch density (${densityMm.toFixed(3)}mm) is sparser than optimal (${VALIDATION_THRESHOLDS.OPTIMAL_DENSITY_MIN}-${VALIDATION_THRESHOLDS.OPTIMAL_DENSITY_MAX}mm)`,
      true,
      "This density is lower than recommended. The design may look sparse or have poor coverage."
    );
  } else {
    return createValidationResult(
      VLevel.DANGER,
      `Stitch density (${densityMm.toFixed(3)}mm) is too sparse. Quality will be poor`,
      true,
      "Stitches are too far apart. The design will have poor coverage and look unprofessional."
    );
  }
}

/**
 * Validate dimensions fit within machine constraints
 */
export function validateDimensions(
  width: number,
  height: number,
  machineProfile: MachineProfile
): ValidationResult {
  if (width > machineProfile.maxHoopWidthMm) {
    return createValidationResult(
      VLevel.CRITICAL,
      `Width (${width.toFixed(1)}mm) exceeds ${machineProfile.name} maximum (${machineProfile.maxHoopWidthMm}mm)`,
      false,
      `The design is too wide for your machine's hoop. Maximum width is ${machineProfile.maxHoopWidthMm}mm.`
    );
  }

  if (height > machineProfile.maxHoopHeightMm) {
    return createValidationResult(
      VLevel.CRITICAL,
      `Height (${height.toFixed(1)}mm) exceeds ${machineProfile.name} maximum (${machineProfile.maxHoopHeightMm}mm)`,
      false,
      `The design is too tall for your machine's hoop. Maximum height is ${machineProfile.maxHoopHeightMm}mm.`
    );
  }

  // Check if close to maximum (warn if within 5mm)
  const widthMargin = machineProfile.maxHoopWidthMm - width;
  const heightMargin = machineProfile.maxHoopHeightMm - height;

  if (widthMargin < 5 || heightMargin < 5) {
    return createValidationResult(
      VLevel.WARNING,
      `Design is close to ${machineProfile.name} hoop limits. Consider adding margin for safety`,
      true,
      `Leaving at least 5mm margin around your design helps with hooping and prevents edge issues.`
    );
  }

  return createValidationResult(
    VLevel.SAFE,
    `Dimensions (${width.toFixed(1)}mm × ${height.toFixed(1)}mm) fit within ${machineProfile.name} hoop`
  );
}

/**
 * Validate stitch length against machine constraints
 */
export function validateStitchLength(
  minLength: number,
  maxLength: number,
  machineProfile: MachineProfile
): ValidationResult {
  if (maxLength > machineProfile.maxStitchLengthMm) {
    return createValidationResult(
      VLevel.DANGER,
      `Maximum stitch length (${maxLength.toFixed(2)}mm) exceeds ${machineProfile.name} limit (${machineProfile.maxStitchLengthMm}mm)`,
      true,
      "Some stitches are longer than recommended. This may cause skipped stitches or poor quality."
    );
  }

  if (minLength < machineProfile.minStitchLengthMm) {
    return createValidationResult(
      VLevel.WARNING,
      `Minimum stitch length (${minLength.toFixed(2)}mm) is below ${machineProfile.name} recommendation (${machineProfile.minStitchLengthMm}mm)`,
      true,
      "Some stitches are very short. This may cause excessive density in some areas."
    );
  }

  return createValidationResult(
    VLevel.SAFE,
    `Stitch lengths are within ${machineProfile.name} acceptable range`
  );
}

/**
 * Run all validations for a resize operation
 *
 * Ported from Python validate_all()
 */
export function validateAll(params: {
  originalWidth: number;
  originalHeight: number;
  newWidth: number;
  newHeight: number;
  newDensityMm?: number;
  minStitchLength?: number;
  maxStitchLength?: number;
  machineProfile: MachineProfile;
}): { canProceed: boolean; results: ValidationResult[] } {
  const {
    originalWidth,
    originalHeight,
    newWidth,
    newHeight,
    newDensityMm,
    minStitchLength,
    maxStitchLength,
    machineProfile,
  } = params;

  const results: ValidationResult[] = [];

  // Check width resize percentage
  const widthResult = validateResizePercentage(originalWidth, newWidth);
  results.push(widthResult);

  // Check height resize percentage
  const heightResult = validateResizePercentage(originalHeight, newHeight);
  results.push(heightResult);

  // Check stitch density if provided
  if (newDensityMm !== undefined) {
    const densityResult = validateStitchDensity(newDensityMm);
    results.push(densityResult);
  }

  // Check dimensions against machine constraints
  const dimensionResult = validateDimensions(newWidth, newHeight, machineProfile);
  results.push(dimensionResult);

  // Check stitch lengths if provided
  if (minStitchLength !== undefined && maxStitchLength !== undefined) {
    const lengthResult = validateStitchLength(minStitchLength, maxStitchLength, machineProfile);
    results.push(lengthResult);
  }

  // Determine if we can proceed (all validations must allow it)
  const canProceed = results.every((r) => r.canProceed);

  return { canProceed, results };
}

/**
 * Get validation summary for display
 */
export function getValidationSummary(results: ValidationResult[]): {
  hasCritical: boolean;
  hasDanger: boolean;
  hasWarning: boolean;
  isSafe: boolean;
  message: string;
} {
  const hasCritical = results.some((r) => r.level === VLevel.CRITICAL);
  const hasDanger = results.some((r) => r.level === VLevel.DANGER);
  const hasWarning = results.some((r) => r.level === VLevel.WARNING);
  const isSafe = results.every((r) => r.level === VLevel.SAFE);

  let message: string;
  if (hasCritical) {
    message = "⛔ CRITICAL: This resize is not recommended!";
  } else if (hasDanger) {
    message = "⚠️  WARNING: Proceed with caution!";
  } else if (hasWarning) {
    message = "⚠️  Some warnings detected. Review before proceeding.";
  } else {
    message = "✅ This resize should produce acceptable results.";
  }

  return { hasCritical, hasDanger, hasWarning, isSafe, message };
}
